File: src/main/java/editor/database/symbol/ColorSymbol.java
package editor.database.symbol;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

import editor.database.attributes.ManaType;

/**
 * This class represents a symbol for a single type of mana.
 *
 * @author Alec Roelke
 */
public class ColorSymbol extends ManaSymbol
{
    /**
     * Map of mana type onto its corresponding ColorSymbol.
     */
    public static final Map<ManaType, ColorSymbol> SYMBOLS = Collections.unmodifiableMap(
            Arrays.stream(ManaType.values()).collect(Collectors.toMap(Function.identity(), ColorSymbol::new)));

    /**
     * Get the symbol corresponding to a color string.
     *
     * @param col String to find the symbol for
     * @return the ColorSymbol corresponding to the given String
     * @throws IllegalArgumentException if the string doesn't correspond to a color symbol
     */
    public static ColorSymbol parseColorSymbol(String col) throws IllegalArgumentException
    {
        return tryParseColorSymbol(col).orElseThrow(() -> new IllegalArgumentException('"' + col + "\" is not a color symbol"));
    }

    /**
     * Get the symbol corresponding to a color string.
     *
     * @param col String to find the symbol for
     * @return the ColorSymbol corresponding to the given String, or null if none exists
     */
    public static Optional<ColorSymbol> tryParseColorSymbol(String col)
    {
        return Optional.ofNullable(SYMBOLS.get(ManaType.tryParseManaType(col)));
    }

    /**
     * Color of this ColorSymbol.
     */
    private final ManaType color;

    /**
     * Create a new ColorSymbol.
     *
     * @param color mana type of the new ColorSymbol.
     */
    private ColorSymbol(ManaType color)
    {
        super(color.toString().toLowerCase() + "_mana.png", String.valueOf(color.shorthand()), 1);
        this.color = color;
    }

    /**
     * {@inheritDoc}
     * This ColorSymbol's weights are 1 for its mana type and 0 for all the others.
     */
    @Override
    public Map<ManaType, Double> colorIntensity()
    {
        return createIntensity(new ColorIntensity(color, 1.0));
    }

    @Override
    public int compareTo(ManaSymbol o)
    {
        if (o instanceof ColorSymbol s)
            return color.compareTo(s.color);
        else
            return super.compareTo(o);
    }
}

File: src/main/scala/editor/database/symbol/ColorSymbol.scala
package editor.database.symbol

import editor.database.attributes.ManaType

import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._

/**
 * Companion to [[ColorSymbol]] containing all (currently) possible mana type symbols, as well as functions for
 * parsing them from strings.
 * @author Alec Roelke
 */
object ColorSymbol {
  @deprecated val SYMBOLS = ColorSymbolGenerator.values.asJava
  @deprecated def tryParseColorSymbol(s: String) = ColorSymbolGenerator.parse(s).toJava
  @deprecated def parseColorSymbol(s: String) = ColorSymbolGenerator.values(ManaType.parseManaType(s))
}

/**
 * Mana symbol representing a specific type of mana. Its intensity is 1 for its mana type, and 0 for all others, as it can only be paid for by that type,
 * and its text representation is just the shorthand of its [[ManaType]].
 * 
 * @constructor create a new symbol for a mana type
 * @param color type of mana the symbol represents
 * 
 * @author Alec Roelke
 */
class ColorSymbol private[symbol](val color: ManaType) extends ManaSymbol(s"${color.toString.toLowerCase}_mana.png", color.shorthand.toString, 1, ColorSymbolGenerator) {
  override def colorIntensity = ManaSymbol.createIntensity(Map(color -> 1))
}